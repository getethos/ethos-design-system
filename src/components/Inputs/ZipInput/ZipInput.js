import React from 'react'
import PropTypes from 'prop-types'
import { TextInput } from '../TextInput';

export const ZipInput = (props) => {
  return (
    <TextInput {...props} />
  )
};

ZipInput.PUBLIC_PROPS = {
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  name: PropTypes.string.isRequired,
  labelCopy: PropTypes.string.isRequired,
  validator: PropTypes.func
}

ZipInput.propTypes = {
  ...ZipInput.PUBLIC_PROPS,
}
