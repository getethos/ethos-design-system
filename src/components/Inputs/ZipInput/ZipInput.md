The `ZipInput` takes a `validator`, and `ZipInputValidator` gives you validation for valid zip format. You can pass that directly, or, wrap it
for any custom or additional logic you may wish to use.
```jsx
import { ZipInputValidator } from './index.js';
<ZipInput
  name="this-zip-input-example"
  allCaps={true}
  labelCopy="What is your zip code?"
  data-tid='the-zip-input'
  validator={(zip) => {
    const errors = ZipInputValidator(zip)
    // More custom logic here should you wish
    return errors
  }}
/>
```
