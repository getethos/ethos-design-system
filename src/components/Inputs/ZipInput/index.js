export { ZipInput } from './ZipInput.js'
