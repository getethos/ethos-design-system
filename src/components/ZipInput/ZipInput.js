import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { TextMaskedInput } from '../TextMaskedInput'
import useErrorMessage from '../../hooks/useErrorMessage.js'
import useInputValidation from '../../hooks/useInputValidation.js'
import zipInputValidator from '../../validators/ZipInputValidator'
import styles from '../TextInput/TextInput.module.scss'
import errorStyles from '../Errors.module.scss'

export const ZipInput = (props) => {
  const {
    name,
    labelCopy,
    allCaps,
    capitalize,
    validator,
    formChangeHandler,
    initialValue,
    currentValue,
    currentError,
    formTouched,
    guide = true,
    keepCharPositions = true,
    autoComplete,
    classOverrides,
    ...restProps
  } = props

  const [getError, setError, , validate] = useErrorMessage(validator)
  const val = currentValue || initialValue
  const [touched, setTouched] = useState(initialValue ? true : false)
  const [value] = useState(val || '')

  // This has to come before useInputValidation setup below
  const callErrorHandlers = (value, handlerFn) => {
    /// Check zip format validity
    let errMsg = zipInputValidator(value)
    const errorMessage = errMsg.length ? errMsg : ''
    if (errorMessage.length) {
      handlerFn(value, errorMessage)
    } else {
      // Call any addtional validators the consumer has setup
      // Note if no validators, we're still safe as validate checks
      let errorMessage = validate(value)
      errorMessage = errorMessage.length ? errorMessage : ''
      handlerFn(value, errorMessage)
    }
  }

  useEffect(() => {
    if (!!formChangeHandler && initialValue) {
      formChangeHandler(initialValue, '')
    }
  }, [])

  const [doValidation] = useInputValidation({
    validate,
    setError,
    formChangeHandler,
    callErrorHandlers,
  })

  const getClasses = () => {
    const base = `ZipInput ${styles.TextInputCommon}`
    return getError(currentError, touched)
      ? `${base} ${errorStyles.Error}`
      : classOverrides
      ? `${base} ${classOverrides}`
      : `${base} ${styles.TextInputStylable}`
  }

  return (
    <>
      <TextMaskedInput
        initialValue={value}
        mask={[/\d/, /\d/, /\d/, /\d/, /\d/]}
        pipe={props.pipe}
        placeholder={props.placeholder}
        type="tel"
        labelCopy={labelCopy}
        allCaps={allCaps}
        capitalize={capitalize}
        data-tid={restProps['data-tid']}
        guide={guide}
        doValidation={doValidation}
        name={name}
        className={getClasses()}
        keepCharPositions={keepCharPositions}
        currentValue={currentValue}
        currentError={currentError}
        formTouched={formTouched}
        setFieldTouched={restProps.setFieldTouched}
        getTouched={touched}
        setTouched={setTouched}
        autoComplete={autoComplete}
        classOverrides={classOverrides}
        labelWeight={props.labelWeight}
        labelColor={props.labelColor}
        disabled={props.disabled}
      />
      {getError(currentError, touched)}
    </>
  )
}

ZipInput.PUBLIC_PROPS = {
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  allCaps: PropTypes.bool,
  /** text transform capitalize label */
  capitalize: PropTypes.bool,
  name: PropTypes.string.isRequired,
  labelCopy: PropTypes.string.isRequired,
  labelWeight: PropTypes.string,
  labelColor: PropTypes.string,
  validator: PropTypes.func,
  initialValue: PropTypes.string,
  guide: PropTypes.bool,
  keepCharPositions: PropTypes.bool,
  autoComplete: PropTypes.string,
}

ZipInput.propTypes = {
  ...ZipInput.PUBLIC_PROPS,
}
