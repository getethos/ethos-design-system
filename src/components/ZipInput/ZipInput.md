The `ZipInput` takes an optional `validator`, but does prevalidatiiin via the `ZipInputValidator` which validates a valid zip format.
```jsx
import { ZipInput } from './index.js';
// formChangeHandler gets wired up automatically if using <Form /> component
const formChangeHandlerStub = () => {}
<ZipInput
  name="this-zip-input-example"
  allCaps={true} 
  labelCopy="What is your zip code?"
  data-tid='the-zip-input'
  formChangeHandler={formChangeHandlerStub}
  validator={(zip) => {
    console.log('ZipInput prevalidates valid zipCode, so we\'d only do this if we wanted to add additional validation')
    return ''
  }}
/>
```

This one sets an `intialValue` which results in the field being considered as
already `touched`. This means you do not have to `blur` for field hint error
messages to appear.

```jsx
import { ZipInput } from './index.js';
const formChangeHandlerStub = () => {}
<ZipInput
  initialValue="94544"
  name="this-zip-input-example2"
  allCaps={true}
  labelCopy="What is your zip code?"
  data-tid='the-zip-input2'
  formChangeHandler={formChangeHandlerStub}
  validator={(zip) => {
    console.log('ZipInput prevalidates valid zipCode, so we\'d only do this if we wanted to add additional validation')
    return ''
  }}
/>
```
