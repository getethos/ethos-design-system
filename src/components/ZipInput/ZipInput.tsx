import * as PropTypes from 'prop-types'
import * as React from 'react'
import { useEffect, useState } from 'react'
import useErrorMessage from '../../hooks/useErrorMessage'
import useInputValidation from '../../hooks/useInputValidation'
import zipInputValidator from '../../validators/ZipInputValidator'
import { PublicFormProps } from '../Form/Form'
import { TextMaskedInput } from '../TextMaskedInput'
const PUBLIC_PROPS = {
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  allCaps: PropTypes.bool,
  /** text transform capitalize label */
  capitalize: PropTypes.bool,
  name: PropTypes.string.isRequired,
  labelCopy: PropTypes.string.isRequired,
  validator: PropTypes.func,
  initialValue: PropTypes.string,
  guide: PropTypes.bool,
  keepCharPositions: PropTypes.bool,
}
type ZipInputProps = PublicFormProps & {
  disabled?: boolean
  allCaps?: boolean
  capitalize?: boolean
  labelCopy: string
  initialValue?: string
  guide?: boolean
  keepCharPositions?: boolean
  placeholder?: string
  pipe?: any
}
export const ZipInput: React.FC<ZipInputProps> & {
  PUBLIC_PROPS: typeof PUBLIC_PROPS
} = (props) => {
  const {
    name,
    labelCopy,
    allCaps,
    capitalize,
    validator,
    formChangeHandler,
    initialValue,
    currentValue,
    currentError,
    formTouched,
    guide = true,
    keepCharPositions = true,
    ...restProps
  } = props
  const [getError, setError, , validate] = useErrorMessage(validator)
  const val = currentValue || initialValue
  const [touched, setTouched] = useState(initialValue ? true : false)
  const [value] = useState(val || '')
  // This has to come before useInputValidation setup below
  const callErrorHandlers = (value, handlerFn) => {
    /// Check zip format validity
    let errMsg = zipInputValidator(value)
    const errorMessage = errMsg.length ? errMsg : ''
    if (errorMessage.length) {
      handlerFn(value, errorMessage)
    } else {
      // Call any addtional validators the consumer has setup
      // Note if no validators, we're still safe as validate checks
      let errorMessage = validate(value)
      errorMessage =
        (typeof errorMessage === 'string' || Array.isArray(errorMessage)) &&
        errorMessage.length
          ? errorMessage
          : ''
      handlerFn(value, errorMessage)
    }
  }
  useEffect(() => {
    if (!!formChangeHandler && initialValue) {
      formChangeHandler(initialValue, '')
    }
  }, [])
  const [doValidation] = useInputValidation({
    validate,
    setError,
    formChangeHandler,
    callErrorHandlers,
  })
  return (
    <>
      <TextMaskedInput
        initialValue={value}
        mask={[/\d/, /\d/, /\d/, /\d/, /\d/]}
        pipe={props.pipe}
        placeholder={props.placeholder}
        type="tel"
        labelCopy={labelCopy}
        allCaps={allCaps}
        capitalize={capitalize}
        data-tid={restProps['data-tid']}
        guide={guide}
        doValidation={doValidation}
        name={name}
        keepCharPositions={keepCharPositions}
        currentValue={currentValue}
        currentError={currentError}
        formTouched={formTouched}
        setFieldTouched={restProps.setFieldTouched}
        getTouched={touched}
        setTouched={setTouched}
      />
      {getError(currentError, touched)}
    </>
  )
}
ZipInput.PUBLIC_PROPS = PUBLIC_PROPS
