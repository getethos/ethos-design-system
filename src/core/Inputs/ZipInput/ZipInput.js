import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { TextMaskedInput } from '../TextMaskedInput'
import useErrorMessage from '../../../hooks/useErrorMessage.js'
import useInputValidation from '../../../hooks/useInputValidation.js'
import zipInputValidator from '../../../validators/ZipInputValidator'
import styles from '../TextInput/TextInput.module.scss'
import errorStyles from '../Errors.module.scss'

export const ZipInput = (props) => {
  const {
    name,
    labelCopy,
    allCaps,
    validator,
    formChangeHandler,
    initialValue,
    currentValue,
    currentError,
    formTouched,
    ...restProps
  } = props

  const [getError, setError, , validate] = useErrorMessage(validator)
  const val = currentValue || initialValue
  const [touched, setTouched] = useState(initialValue ? true : false)
  const [value] = useState(val || '')

  // This has to come before useInputValidation setup below
  const callErrorHandlers = (value, handlerFn) => {
    /// Check zip format validity
    let errMsg = zipInputValidator(value)
    const errorMessage = errMsg.length ? errMsg : ''
    if (errorMessage.length) {
      handlerFn(value, errorMessage)
    } else {
      // Call any addtional validators the consumer has setup
      // Note if no validators, we're still safe as validate checks
      let errorMessage = validate(value)
      errorMessage = errorMessage.length ? errorMessage : ''
      handlerFn(value, errorMessage)
    }
  }

  useEffect(() => {
    if (!!formChangeHandler && initialValue) {
      console.log('ZipInput useEffect -- calling doValidation')
      formChangeHandler(initialValue, '')
    }
  }, [])

  const [doValidation] = useInputValidation({
    validate,
    setError,
    formChangeHandler,
    callErrorHandlers,
  })

  const getClasses = () => {
    return getError(currentError, touched)
      ? `ZipInput ${styles.TextInput} ${errorStyles.Error}`
      : `ZipInput ${styles.TextInput}`
  }

  return (
    <>
      <TextMaskedInput
        initialValue={value}
        mask={[/\d/, /\d/, /\d/, /\d/, /\d/]}
        pipe={props.pipe}
        placeholder={props.placeholder}
        type="tel"
        labelCopy={labelCopy}
        allCaps={allCaps}
        data-tid={restProps['data-tid']}
        guide={true}
        doValidation={doValidation}
        name={name}
        className={getClasses()}
        keepCharPositions={true}
        currentValue={currentValue}
        currentError={currentError}
        formTouched={formTouched}
        setFieldTouched={restProps.setFieldTouched}
        getTouched={touched}
        setTouched={setTouched}
      />
      {getError(currentError, touched)}
    </>
  )
}

ZipInput.PUBLIC_PROPS = {
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  allCaps: PropTypes.bool,
  name: PropTypes.string.isRequired,
  labelCopy: PropTypes.string.isRequired,
  validator: PropTypes.func,
  initialValue: PropTypes.string,
}

ZipInput.propTypes = {
  ...ZipInput.PUBLIC_PROPS,
}
