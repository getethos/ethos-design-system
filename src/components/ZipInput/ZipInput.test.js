import React from 'react'
import { ZipInput } from './ZipInput'
import renderer from 'react-test-renderer'

// TODO: make a helper
// https://github.com/text-mask/text-mask/issues/427
jest.mock('react-text-mask', () => (props) => <input type="text" {...props} />)

describe('ZipInput', () => {
  it('default rendering', () => {
    const tree = renderer
      .create(
        <ZipInput
          name="le-zip"
          allCaps={true}
          labelCopy="What is your zip code?"
          data-tid="le-zip"
          validator={() => {}}
          autoComplete="postal-code"
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
