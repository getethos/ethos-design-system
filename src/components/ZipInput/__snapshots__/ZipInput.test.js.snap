// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZipInput default rendering 1`] = `
Array [
  <label
    className="Caption Theinhardt Medium500 GrayPrimary AllCaps"
    htmlFor="le-zip"
  >
    What is your zip code?
  </label>,
  <div
    aria-hidden={true}
    className="Spacer"
    style={
      Object {
        "height": 4,
      }
    }
  />,
  <input
    className="TextMaskedInput TextInput"
    data-tid="le-zip"
    disabled={false}
    guide={true}
    keepCharPositions={true}
    mask={
      Array [
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
      ]
    }
    name="le-zip"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="tel"
    value=""
  />,
  "",
]
`;
